sw/device/lib/testing/test_framework/freertos_port.c

```diff
@@ -0,0 +1,101 @@
+#include "sw/vendor/freertos_freertos_kernel/include/task.h"
+#include "sw/vendor/freertos_freertos_kernel/portable/GCC/RISC-V/portmacro.h"
+
+// TODO: make this toplevel agnostic.
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+// NOTE: some of the function names below do NOT, and cannot, conform to the
+// style guide, since they are specific implementations of FreeRTOS defined
+// functions.
+
+// ----------------------------------------------------------------------------
+// Timer.
+// ----------------------------------------------------------------------------
+static dif_rv_timer_t timer;
+static const uint32_t kTimerHartId = (uint32_t)kTopEarlgreyPlicTargetIbex0;
+static const uint32_t kTimerComparatorId = 0;
+static const uint64_t kTimerDeadline =
+    100;  // Counter must reach 100 for an IRQ to be triggered.
+
+void vPortSetupTimerInterrupt(void) {
+  LOG_INFO("Configuring timer interrupt ...");
+
+  // Initialize and reset the timer.
+  mmio_region_t timer_reg =
+      mmio_region_from_addr(TOP_EARLGREY_RV_TIMER_BASE_ADDR);
+  CHECK(dif_rv_timer_init(
+            timer_reg,
+            (dif_rv_timer_config_t){.hart_count = 1, .comparator_count = 1},
+            &timer) == kDifRvTimerOk);
+
+  // Compute and set tick parameters (i.e., step, prescale, etc.).
+  dif_rv_timer_tick_params_t tick_params;
+  CHECK(dif_rv_timer_approximate_tick_params(
+            kClockFreqPeripheralHz, configTICK_RATE_HZ * kTimerDeadline,
+            &tick_params) == kDifRvTimerApproximateTickParamsOk);
+  LOG_INFO("Tick Freq. (Hz): %u, Prescale: %u; Tick Step: %u",
+           (uint32_t)kClockFreqPeripheralHz, (uint32_t)tick_params.prescale,
+           (uint32_t)tick_params.tick_step);
+  CHECK(dif_rv_timer_set_tick_params(&timer, kTimerHartId, tick_params) ==
+        kDifRvTimerOk);
+
+  // Enable RV Timer interrupts and arm/enable the timer.
+  CHECK(dif_rv_timer_irq_enable(&timer, kTimerHartId, kTimerComparatorId,
+                                kDifRvTimerEnabled) == kDifRvTimerOk);
+  CHECK(dif_rv_timer_arm(&timer, kTimerHartId, kTimerComparatorId,
+                         kTimerDeadline) == kDifRvTimerOk);
+
+  CHECK(dif_rv_timer_counter_set_enabled(&timer, kTimerHartId,
+                                         kDifRvTimerEnabled) == kDifRvTimerOk);
+}
+
+// ----------------------------------------------------------------------------
+// Scheduler.
+// ----------------------------------------------------------------------------
+extern void xPortStartFirstTask(void);
+
+BaseType_t xPortStartScheduler(void) {
+  vPortSetupTimerInterrupt();
+  irq_timer_ctrl(true);
+  irq_external_ctrl(true);
+  irq_software_ctrl(true);
+  xPortStartFirstTask();
+
+  // Unreachable.
+  return pdFAIL;
+}
+
+void vPortEndScheduler(void) {
+  // Not implemented.
+  // TODO: trigger this to be called when from the idle task hook when all tests
+  // have completed.
+  for (;;)
```

Comment:
Should be a while(true) w wait_for_interrupt. Make sure you catch this everywhere else!

Reply:
fixed.